### PR TITLE
feature: CCV Channel-based Message Filtering

### PR DESCRIPTION
**Pull request by alexanderbez**
Add an AnteHandler that does message filtering based on if the CCV channel has been established or not. If it has been established, the AnteHandler decorator performs a no-op, i.e. all messages are OK. Otherwise, only certain IBC messages are supported.

closes: #74

### CODE REVIEW

File: app/consumer/msg_filter_ante.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+		}
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+func hasValidMsgs(msgs []sdk.Msg) bool {
+	for _, msg := range msgs {
+		msgType := sdk.MsgTypeURL(msg)
+		// TODO: Perform message filtering that only allows certain IBC messages
```
Comment (alexanderbez) [inline app/consumer/msg_filter_ante.go:40]: @mpoke do you know what message types we need to support when the CCV channel has not been established yet?

Comment (mpoke) [reply app/consumer/msg_filter_ante.go:42]: @alexanderbez All IBC messages. The client needs to be updated, the connection and the channel need to be set up, and then a normal IBC packet will arrive with valset updates, which will mark the CCV channel as established.

Besides, I don't see any other message that is necessary during channel initialization.

**Comment by mpoke** [comment]
@alexanderbez Issue https://github.com/cosmos/interchain-security/issues/115 is related

File: app/consumer/ante/msg_filter_ante.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+func hasValidMsgsPreCCV(msgs []sdk.Msg) bool {
+	for _, msg := range msgs {
+		msgType := sdk.MsgTypeURL(msg)
+
+		// Only accept IBC messages prior to the CCV channel being established.
+		// Note, rather than listing out all possible IBC message types, we assume
+		// all IBC message types have a correct and canonical prefix -- /ibc.*
+		if !strings.HasPrefix(msgType, "/ibc.") {
```
Comment (AdityaSripal) [inline app/consumer/ante/msg_filter_ante.go:52]: This seems good to me. I don't think we need to filter specific core IBC messages out per se. What do you think @mpoke ?

**Review by AdityaSripal — APPROVED** [review]
Thanks for the addition @alexanderbez !

File: app/consumer/ante/msg_filter_ante.go (hunk 2 above)
Comment (mpoke) [reply app/consumer/ante/msg_filter_ante.go:52]: I agree. I think all IBC messages should be enabled at all times.

File: app/consumer/ante/msg_filter_ante.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+func NewMsgFilterDecorator(k ConsumerKeeper) MsgFilterDecorator {
+	return MsgFilterDecorator{
+		ConsumerKeeper: k,
+	}
+}
+
+func (mfd MsgFilterDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	currHeight := ctx.BlockHeight()
+
+	// If the CCV channel has yet been established, then we must only allow certain
```
Comment (mpoke) [inline app/consumer/ante/msg_filter_ante.go:34]:
```suggestion
	// If the CCV channel has not yet been established, then we must only allow certain
```

**Review by mpoke — APPROVED** [review]
LGTM. Nice work @alexanderbez

**Comment by mpoke** [comment]
@alexanderbez Is this ready to merge or there is something else required? The TODO in the PR's description seems to be completed.

**Pull request merged**